libs/filer/workspace_files_extensions_client.go

```diff
@@ -0,0 +1,328 @@
+// The "repos_export_format" field is not exposed by the SDK.
+type workspaceFileStatus struct {
+	*workspace.ObjectInfo
+
+	// The export format of the notebook. This is not exposed by the SDK.
+	ReposExportFormat workspace.ExportFormat `json:"repos_export_format,omitempty"`
+
+	// Name of the file to be used in any API calls made using the workspace files
+	// filer. For notebooks this path does not include the extension.
+	nameForWorkspaceAPI string
+}
+
+// A custom unmarsaller for the workspaceFileStatus struct. This is needed because
+// workspaceFileStatus embeds the workspace.ObjectInfo which itself has a custom
+// unmarshaller.
+// If a custom unmarshaller is not provided extra fields like ReposExportFormat
+// will not have values set.
+func (s *workspaceFileStatus) UnmarshalJSON(b []byte) error {
+	return marshal.Unmarshal(b, s)
+}
+
+func (s *workspaceFileStatus) MarshalJSON() ([]byte, error) {
+	return marshal.Marshal(s)
+}
+
+func (w *workspaceFilesExtensionsClient) stat(ctx context.Context, name string) (*workspaceFileStatus, error) {
+	stat := &workspaceFileStatus{
+		nameForWorkspaceAPI: name,
+	}
+
+	// Perform bespoke API call because "return_export_info" is not exposed by the SDK.
+	// We need "repos_export_format" to determine if the file is a py or a ipynb notebook.
+	// This is not exposed by the SDK so we need to make a direct API call.
+	err := w.apiClient.Do(
+		ctx,
+		http.MethodGet,
+		"/api/2.0/workspace/get-status",
+		nil,
+		map[string]string{
+			"path":               path.Join(w.root, name),
+			"return_export_info": "true",
+		},
+		stat,
+	)
+	return stat, err
+}
+
+// This function returns the stat for the provided notebook. The stat object itself
+// contains the path with the extension since it is meant used in the context of a fs.FileInfo.
+func (w *workspaceFilesExtensionsClient) removeNotebookExtension(ctx context.Context, name string) (stat *workspaceFileStatus, ok bool) {
+	ext := path.Ext(name)
+
+	nameWithoutExtension := strings.TrimSuffix(name, ext)
+
+	// File name does not have an extension associated that is with Databricks
+	// notebooks, return early.
+	if _, ok := extensionsToLanguages[ext]; !ok {
+		return nil, false
+	}
+
+	// If the file could be a notebook, check if it is and has the correct language.
+	stat, err := w.stat(ctx, nameWithoutExtension)
+	if err != nil {
+		log.Debugf(ctx, "attempting to determine if %s could be a notebook. Failed to fetch the status of object at %s: %s", name, path.Join(w.root, nameWithoutExtension), err)
+		return nil, false
+	}
+
+	// Not a notebook. Return early.
+	if stat.ObjectType != workspace.ObjectTypeNotebook {
+		log.Debugf(ctx, "attempting to determine if %s could be a notebook. Found an object at %s but it is not a notebook. It is a %s.", name, path.Join(w.root, nameWithoutExtension), stat.ObjectType)
+		return nil, false
+	}
+
+	// Not the correct language. Return early.
+	if stat.Language != extensionsToLanguages[ext] {
```

Comment:
Out of curiosity, how can this happen?

Reply:
Let's say we're trying to stat `foo.scala`. Without extension this is `foo`. It can be a Python notebook and the language returned in the stat is `PYTHON`. Then this check will return false, because we can't (and shouldn't) access `foo.scala` if it is a Python notebook, only if it is a Scala notebook.